pallets/permissions/src/lib.rs

```diff
@@ -216,51 +223,56 @@ impl<T: Config> Pallet<T> {
 		who: T::AccountId,
 		role: T::Role,
 	) -> Result<(), DispatchError> {
-		Permission::<T>::try_get(who.clone(), location.clone()).map_or(
-			{
-				let mut new_role = T::Storage::default();
-				new_role
-					.add(role.clone())
-					.map_err(|_| Error::<T>::WrongParameters)?;
-
-				Permission::<T>::insert(who.clone(), location.clone(), new_role);
-				Ok(())
-			},
-			|mut roles| {
-				if !roles.exists(role.clone()) {
-					roles.add(role).map_err(|_| Error::<T>::WrongParameters)?;
-
-					Permission::<T>::insert(who.clone(), location, roles);
-					Ok(())
-				} else {
+		PermissionCount::<T>::try_mutate(location.clone(), |perm_count| {
+			let num_permissions = perm_count.map_or(1, |count| count + 1);
+			if num_permissions > T::MaxRolesPerLocation::get() {
+				return Err(Error::<T>::TooManyRoles.into());
+			}
+			*perm_count = Some(num_permissions);
+
+			Permission::<T>::try_mutate(who.clone(), location.clone(), |maybe_roles| {
+				let mut roles = maybe_roles.take().unwrap_or_default();
+				if roles.exists(role.clone()) {
 					Err(Error::<T>::RoleAlreadyGiven.into())
+				} else {
+					roles
+						.add(role.clone())
+						.map_err(|_| Error::<T>::WrongParameters)?;
+					*maybe_roles = Some(roles);
+					Ok(())
 				}
-			},
-		)
+			})
+		})
 	}
 
 	fn do_rm_permission(
 		location: T::Location,
 		who: T::AccountId,
 		role: T::Role,
 	) -> Result<(), DispatchError> {
-		Permission::<T>::try_get(who.clone(), location.clone()).map_or(
-			Err(Error::<T>::NoRoles.into()),
-			|mut roles| {
+		PermissionCount::<T>::try_mutate(location.clone(), |perm_count| {
+			let num_permissions = perm_count.map_or(0, |count| count - 1);
+			if num_permissions == 0 {
+				*perm_count = None;
+			} else {
+				*perm_count = Some(num_permissions);
+			}
+
+			Permission::<T>::try_mutate(who.clone(), location.clone(), |maybe_roles| {
+				let mut roles = maybe_roles.take().ok_or(Error::<T>::NoRoles)?;
 				if roles.exists(role.clone()) {
 					roles.rm(role).map_err(|_| Error::<T>::WrongParameters)?;
-
 					if roles.empty() {
-						Permission::<T>::remove(who, location);
+						*maybe_roles = None
 					} else {
-						Permission::<T>::insert(who, location, roles);
+						*maybe_roles = Some(roles)
```

Comment:
As above. Could we add a test that checks if the `num_permissions` is decreased correctly and both counts and roles are removed from storage ?